File: chapt10/chapt10_ws/src/learn_qos_py/learn_qos_py/reliability_test.py
import rclpy
from rclpy.node import Node
from nav_msgs.msg import Odometry
from rclpy import qos
from rclpy.duration import Duration



class OdomPublisherSubscriber(Node):

    def __init__(self):
        super().__init__('odom_publisher_subscriber')
        # 创建发布者并设置QoS为sensor
        qos_profile = qos.QoSProfile(
            depth=10,  # 队列深度
            reliability=qos.ReliabilityPolicy.BEST_EFFORT,  # 可靠性
            durability=qos.DurabilityPolicy.TRANSIENT_LOCAL,  # 持久性
            history=qos.HistoryPolicy.KEEP_LAST,  # 历史记录策略
            deadline=Duration(seconds=1.0, nanoseconds=0),
        )
        # 在订阅和发布中使用
        self.odom_publisher = self.create_publisher(Odometry, 'odom', qos_profile)
        # 创建订阅者（默认QoS配置）队列深度设置为 5
        self.odom_subscriber = self.create_subscription(
            Odometry, 'odom', self.odom_callback,qos.qos_profile_sensor_data)
        # 创建一个1秒的定时器，并指定回调函数
        self.timer = self.create_timer(1.0, self.timer_callback)

    def odom_callback(self, msg):
        self.get_logger().info('收到里程计消息')

    def timer_callback(self):
        odom_msg = Odometry()  # 创建一个Odometry消息
        self.odom_publisher.publish(odom_msg)  # 发布消息


def main(args=None):
    rclpy.init(args=args)
    odom_node = OdomPublisherSubscriber()
    rclpy.spin(odom_node)
    rclpy.shutdown()

File: chapt10/chapt10_ws/src/learn_qos_cpp/src/reliability_test.cpp
#include <nav_msgs/msg/odometry.hpp>
#include <rclcpp/rclcpp.hpp>

class OdomPublisherSubscriber : public rclcpp::Node
{
public:
    OdomPublisherSubscriber() : Node("odom_publisher_subscriber")
    {
        rclcpp::QoS qos_profile(10);                                       // 队列深度为10
        qos_profile.reliability(RMW_QOS_POLICY_RELIABILITY_BEST_EFFORT);   // 可靠性策略
        qos_profile.durability(RMW_QOS_POLICY_DURABILITY_TRANSIENT_LOCAL); // 持久性策略
        qos_profile.history(RMW_QOS_POLICY_HISTORY_KEEP_LAST);             // 历史记录策略
        qos_profile.deadline(rclcpp::Duration(1, 0));                      // 截止时间为1秒

        odom_publisher_ = this->create_publisher<nav_msgs::msg::Odometry>(
            "odom", qos_profile);

        // // 创建发布者并设置QoS为sensor
        // odom_publisher_ = this->create_publisher<nav_msgs::msg::Odometry>(
        //     "odom", rclcpp::SensorDataQoS());

        // 创建订阅者（默认QoS配置）队列深度设置为 5
        // odom_subscription_ = this->create_subscription<nav_msgs::msg::Odometry>(
        //     "odom", 5,
        //     [this](const nav_msgs::msg::Odometry::SharedPtr msg) {
        //       (void)msg;
        //       RCLCPP_INFO(this->get_logger(), "收到里程计消息");
        //     });

        odom_subscription_ = this->create_subscription<nav_msgs::msg::Odometry>(
            "odom", rclcpp::SensorDataQoS(),
            [this](const nav_msgs::msg::Odometry::SharedPtr msg)
            {
                (void)msg;
                RCLCPP_INFO(this->get_logger(), "收到里程计消息");
            });

        // 创建一个1秒的定时器，并指定回调函数
        timer_ = this->create_wall_timer(std::chrono::seconds(1), [this]()
                                         { odom_publisher_->publish(nav_msgs::msg::Odometry()); });
    }

private:
    rclcpp::Publisher<nav_msgs::msg::Odometry>::SharedPtr odom_publisher_;
    rclcpp::Subscription<nav_msgs::msg::Odometry>::SharedPtr odom_subscription_;
    rclcpp::TimerBase::SharedPtr timer_;
};

int main(int argc, char *argv[])
{
    rclcpp::init(argc, argv);
    auto odom_node = std::make_shared<OdomPublisherSubscriber>();
    rclcpp::spin(odom_node);
    rclcpp::shutdown();
    return 0;
}